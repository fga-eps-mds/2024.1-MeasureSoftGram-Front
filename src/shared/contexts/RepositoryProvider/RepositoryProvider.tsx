import React, { createContext, useState, useContext, ReactNode, useMemo } from 'react';

import { Repository, Historical, Result } from '@customTypes/repository';
import { RepositoriesLatestTsqmi, TsqmiValue } from '@customTypes/product';

interface Props {
  children: ReactNode;
}

interface IRepositoryContext {
  currentRepository?: Repository;
  setCurrentRepository: (repository: Repository) => void;
  repositoryList?: Repository[];
  setRepositoryList: (repository: Repository) => void;
  characteristics: string[];
  setCharacteristics: (characteristics: string[]) => void;
  subCharacteristics: string[];
  setSubCharacteristics: (subCharacteristics: string[]) => void;
  measures: string[];
  setMeasures: (measures: string[]) => void;
  metrics: string[];
  setMetrics: (metrics: string[]) => void;
  historicalTSQMI: Historical;
  setHistoricalTSQMI: (historical: Historical) => void;
  latestTSQMI: TsqmiValue;
  setLatestTSQMI: (result: TsqmiValue) => void;
  latestTSQMIBadgeUrl: string;
  setLatestTSQMIBadgeUrl: (result: string) => void;
  repositoriesLatestTsqmi: RepositoriesLatestTsqmi;
  setRepositoriesLatestTsqmi: (result: RepositoriesLatestTsqmi) => void;
}

const RepositoryContext = createContext<IRepositoryContext | undefined>(undefined);

export function RepositoryProvider({ children }: Props) {
  const [currentRepository, setCurrentRepository] = useState<Repository | undefined>();
  const [repositoryList, setRepositoryList] = useState<Repository[]>();

  const [characteristics, setCharacteristics] = useState<string[]>([]);
  const [subCharacteristics, setSubCharacteristics] = useState<string[]>([]);
  const [measures, setMeasures] = useState<string[]>([]);
  const [metrics, setMetrics] = useState<string[]>([]);
  const [historicalTSQMI, setHistoricalTSQMI] = useState<Historical>();
  const [latestTSQMI, setLatestTSQMI] = useState<TsqmiValue>();
  const [latestTSQMIBadgeUrl, setLatestTSQMIBadgeUrl] = useState<string>();
  const [repositoriesLatestTsqmi, setRepositoriesLatestTsqmi] = useState<RepositoriesLatestTsqmi>();


  const value = useMemo(
    () => ({
      currentRepository,
      setCurrentRepository,
      repositoryList,
      setRepositoryList,
      characteristics,
      setCharacteristics,
      subCharacteristics,
      setSubCharacteristics,
      measures,
      setMeasures,
      metrics,
      setMetrics,
      historicalTSQMI,
      setHistoricalTSQMI,
      latestTSQMI,
      setLatestTSQMI,
      latestTSQMIBadgeUrl,
      setLatestTSQMIBadgeUrl,
      repositoriesLatestTsqmi,
      setRepositoriesLatestTsqmi
    }),
    [currentRepository, repositoryList, characteristics, subCharacteristics, measures, metrics, historicalTSQMI, latestTSQMI]
  );

  return <RepositoryContext.Provider value={value}>{children}</RepositoryContext.Provider>;
}

export function useRepositoryContext() {
  const context = useContext(RepositoryContext);

  if (context === undefined) {
    throw new Error('useRepositoryContext must be used within a RepositoryContext');
  }

  return context;
}
