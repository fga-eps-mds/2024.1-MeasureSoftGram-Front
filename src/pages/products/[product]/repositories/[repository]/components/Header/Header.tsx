import { useRepositoryContext } from '@contexts/RepositoryProvider';
import { Alert, Box, Button, IconButton, Modal, Typography } from '@mui/material';
import WarningIcon from '@mui/icons-material/Warning';
import SettingsIcon from '@mui/icons-material/Settings';

import React, { useEffect, useState } from 'react';
import { toast } from 'react-toastify';
import { useRouter } from 'next/router';
import { getPathId } from '@utils/pathDestructer';
import { useProductContext } from '@contexts/ProductProvider';
import ReactEcharts from 'echarts-for-react';
import { ProductFormData } from '@services/product';
import { toNumber } from 'lodash';
import { useProductQuery } from '@pages/products/hooks/useProductQuery';
import GaugeSlider from '../GaugeSlider';
import CopyBadgeModal from '../CopyBadgeModal';

function Header() {
  const { currentProduct, setCurrentProduct } = useProductContext();
  const { currentRepository } = useRepositoryContext();
  const { updateProduct } = useProductQuery();

  const [openModal, setOpenModal] = useState<boolean>(false);

  const initialValues = currentProduct && [currentProduct.gaugeRedLimit, currentProduct.gaugeYellowLimit];
  const [values, setValues] = useState(
    initialValues
  );

  const { query } = useRouter();

  useEffect(() => {
    if (initialValues && !values) {
      setValues(initialValues);
    }
  }, [initialValues]);

  const handleOpenModal = () => {
    setOpenModal(true);
  };

  const handleCloseModal = () => {
    setValues(initialValues);
    setOpenModal(false);
  };

  const option = {
    series: {
      type: 'gauge',
      startAngle: 180,
      endAngle: 360,
      min: 0,
      max: 1,
      axisLine: {
        lineStyle: {
          width: 50,
          color: [
            [(values ? values[0] : 0.33), '#e74c3c'],
            [(values ? values[1] : 0.66), '#f1c40f'],
            [1, '#07bc0c']
          ]
        }
      },
      axisTick: {
        length: 0
      },
      splitLine: {
        length: 0
      },
      axisLabel: {
        distance: 0,
        rotate: 'tangential',
        formatter() {
          return '';
        }
      }
    }
  };

  const onSubmit = async () => {
    if (values && currentProduct) {
      const [organizationId, productId] = getPathId(query?.product as string);
      const productData: ProductFormData = {
        name: currentProduct.name,
        gaugeRedLimit: values[0],
        gaugeYellowLimit: values[1],
        organizationId: toNumber(organizationId)
      }

      const result = await updateProduct(productId, productData);
      if (result.type === 'success') {
        toast.success('Intervalos atualizados com sucesso!');
        setCurrentProduct(result.value);
      } else {
        toast.success('Erro ao atualizar intervalos!');
      }

      setOpenModal(false);
    }
  };

  return (
    <Box display="flex" flexDirection="row" justifyContent="space-between">
      <Box display="flex" flexDirection="column">
        <Box display="flex">
          <Typography variant="h4" marginRight="10px">
            Repositório
          </Typography>
          <Typography variant="h4" fontWeight="500" color="#33568E">
            {currentRepository?.name}
          </Typography>
        </Box>
        <Typography variant="caption" color="gray">
          {currentRepository?.description}
        </Typography>
      </Box>
      <Box>
        <IconButton onClick={handleOpenModal}>
          <SettingsIcon />
        </IconButton>
        <CopyBadgeModal />
      </Box>

      <Modal
        open={openModal}
        onClose={handleCloseModal}
      >
        <Box
          sx={{
            width: 550,
            p: 2,
            position: 'absolute',
            top: '50%', left: '50%',
            transform: 'translate(-50%, -50%)',
            bgcolor: 'background.paper',
            boxShadow: 24,
            borderRadius: 1,
          }}
        >
          <Typography variant="h6" gutterBottom>
            Editar Intervalos
          </Typography>
          {values ?
            <>
              <Box
                style={{
                  height: '180px',
                }}
              >
                <ReactEcharts option={option} />
              </Box>
              <Box>
                <GaugeSlider
                  step={0.01}
                  initialValues={values}
                  min={0}
                  max={1}
                  values={values}
                  setValues={setValues}
                />
              </Box>
              <Box
                display="flex"
                flexDirection="row"
                justifyContent="center"
                gap="20px"
                marginBottom='10px'
              >
                <Button
                  onClick={handleCloseModal}
                  variant='outlined'
                >
                  Cancelar
                </Button>
                <Button
                  variant='contained'
                  onClick={onSubmit}
                >
                  Salvar
                </Button>
              </Box>
              <Alert icon={<WarningIcon />} sx={{ display: "flex", justifyContent: "center", textAlign: 'center' }} severity="warning">
                Atenção: Essa configuração será aplicada a todos os repositórios do produto.
              </Alert>
            </>
            :
            <Alert sx={{ display: "flex", justifyContent: "center", textAlign: 'center' }} severity="error">
              Ocorreu um erro ao tentar carregar as informações.
            </Alert>
          }
        </Box>
      </Modal>
    </Box>
  );
}

export default Header;
