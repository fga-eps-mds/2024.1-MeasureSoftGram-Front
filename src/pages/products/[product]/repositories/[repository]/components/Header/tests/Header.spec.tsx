import React from 'react';
import { RepositoryProvider } from '@contexts/RepositoryProvider';
import { render, fireEvent } from '@testing-library/react';
import Header from '../Header';

interface Props {
  children: React.ReactNode;
}

jest.mock('@contexts/OrganizationProvider', () => ({
  useOrganizationContext: jest.fn(() => ({
    currentOrganization: {
      id: '1'
    }
  }))
}));

jest.mock('@contexts/ProductProvider', () => ({
  useProductContext: jest.fn(() => ({
    currentProduct: {
      id: '1',
      name: 'MeasureSoftGram',
      gaugeRedLimit: 0.33,
      gaugeYellowLimit: 0.66
    }
  }))
}));


const AllTheProviders = ({ children }: Props) => (

  <RepositoryProvider>{children}</RepositoryProvider>

);

describe('Header', () => {
  it('should render correctly', () => {
    const { container } = render(<Header />, {
      wrapper: AllTheProviders
    });

    expect(container).toMatchSnapshot();
  });

  it('opens and closes the modal correctly', () => {
    const { getByText, queryByText, getByTestId } = render(<Header />, { wrapper: AllTheProviders });

    const testString = 'Editar Intervalos';

    expect(queryByText(testString)).toBeFalsy();

    const settingsButton = getByTestId('SettingsIcon');
    expect(settingsButton).toBeTruthy();

    fireEvent.click(settingsButton);
    expect(queryByText(testString)).toBeTruthy();

    const cancelButton = getByText('Cancelar');
    expect(cancelButton).toBeTruthy();
    fireEvent.click(cancelButton);
    expect(queryByText(testString)).toBeFalsy();

  });

  it('initialValues should be 0.33 and 0.66', () => {
    const { queryByText, getByTestId } = render(<Header />, { wrapper: AllTheProviders });

    expect(queryByText('Editar Intervalos')).toBeFalsy();

    const settingsButton = getByTestId('SettingsIcon');
    expect(settingsButton).toBeTruthy();

    fireEvent.click(settingsButton);
    expect(queryByText('0.33')).toBeTruthy();
    expect(queryByText('0.66')).toBeTruthy();

  });

});
