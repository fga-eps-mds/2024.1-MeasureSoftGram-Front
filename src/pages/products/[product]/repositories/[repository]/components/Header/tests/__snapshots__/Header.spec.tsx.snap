// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render correctly 1`] = `
<div>
  <div
    class="MuiBox-root css-9jay18"
  >
    <div
      class="MuiBox-root css-j7qwjs"
    >
      <div
        class="MuiBox-root css-k008qs"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 css-obdthj-MuiTypography-root"
        >
          repository
        </h4>
        <h4
          class="MuiTypography-root MuiTypography-h4 css-r2wbd1-MuiTypography-root"
        />
      </div>
      <span
        class="MuiTypography-root MuiTypography-caption css-qkcnm0-MuiTypography-root"
      />
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="SettingsIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61l-2.01-1.58zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ContentCopyIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2zm0 16H8V7h11v14z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
